Add replacement implementation of Lua IO

wingo opened the pull request:
The goal is to replace Lua's "io" library with a compatible interface, except written in Lua instead of C, and allowing extra functionality such as string files, mmap buffers, and non-blocking concurrency.

Eventually this will allow for a number of refactorings:

 * `lib.ctable` will be able to use these standard streams for its serialization and deserialization
 * `lib.yang.binary` likewise
 * `lib.yang.parser` will be able to take a stream as its argument instead of a string
 * `lib.yang.data` too, we get to remove `lib.yang.stream`
 * `lib.ptree.json` likewise
 * we'll be able to use fibers when needed, perhaps in the `lib.ptree` manager, simplifying parsing and serialization of RPCs; the stream abstraction is ready for cooperative coroutine-based multitasking and integration with an epoll loop
 * `snabb config` will be able to stream input and output, avoiding the need to build enormous strings for large configurations (i.e. large routing tables)

I think in the end we will have more functionality with less code and still good speed.

The current status is that the `file` methods are complete but the `io` module is not yet done.

wingo committed:
Add ring buffer module

dpino commented on src/lib/buffer.lua:
```diff
@@ -0,0 +1,161 @@
+end
+
+function buffer:drop()
+   if count > self:read_avail() then error('read xrun') end
```
`count` is not defined.

dpino reviewed (commented):
Looking good!

wingo committed:
Add full replacement for Lua's "io" module

To install, run `require('lib.stream.compat').install()`.

wingo committed:
Add basic IO streams

The goal is to replace Lua's "io" library with a compatible interface,
except written in Lua instead of C, and allowing extra functionality
such as string files, mmap buffers, and non-blocking concurrency.

This commit adds the common stream functionality, which corresponds to a
superset of the interface of Lua files, as implemented by LuaJIT.  Note
that some functionality is present but not implemented.

wingo committed:
Add file stream implementation

wingo commented:
I rebased this commit to be based on `next` instead of `lwaftr`, and fed in some feedback from fibers, which I also rebased on top of this work.  Apologies for any confusion!

wingo committed:
Move io.open implementation to lib.stream.file

Also add optional "perms" argument.

wingo committed:
YANG output byte streams use new Lua streams

wingo committed:
Move tmpfile() impl to file.lua; add optional perms arg

wingo committed:
More YANG output byte stream refactoring

wingo committed:
Remove calls to :align() output to ffi natural sizes

As we won't mmap and cast directly (instead we'll copy through the
stream interface), the alignment is useless.

wingo committed:
Add some helper methods to Lua streams

wingo committed:
Remove unused open_output_byte_stream function

wingo committed:
Replace lib.yang streams with lib.stream streams

Adapts lib.yang and lib.ctable loading and saving to use the new
lib.stream I/O implementation.

wingo committed:
Make "offset" arg to lseek a signed integer

Otherwise, passing a negative seek amount as a normal Lua number will
involve a cast from double to uint64.  In C it's undefined behavior when
a double outside the [0,2^64) range is cast to uint64.  In Lua we try to
additionally accomodate the range [-2^63, -1], but there is a bug on
x64-64 and might be a bug on other platforms:

  https://github.com/LuaJIT/LuaJIT/pull/415

However it's cheaper to simply target an int64_t when you want to allow
negative numbers, as is our case, because you don't have to do assembly
heroics for it to do what you want.  The `tonumber` call in the return
of `linux.lua:lseek` indicates anyway that our range is not the full
64-bit range, so we might as well restrict instead to long rather than
ulong.

wingo committed:
Remove unused lwaftr rangemap load/save capabilities

wingo committed:
Remove YANG streams

wingo committed:
Add memory-backed stream implementation

wingo committed:
Migrate YANG and related libs away from ad-hoc string_io_file

Instead, they use full-fledged streams, though still backed by memory.

wingo committed:
YANG syntax parser is now streaming

This commit makes the YANG parser accept a stream instead of a string,
and adapts all eventual callers to work in streams instead of strings,
creating temporary streams as needed.

Eventually the RPC protocol between the ptree manager and the "snabb
config" client should also be changed to allow streaming (not requiring
the config to be written out as a string).

wingo commented:
After a couple more rebases, this is ready to go.

Everything is streaming except the RPC protocol, which has a couple of cases in which it serializes configurations as strings.  One is in the YANG side:

```yang
  rpc get-schema {
    input {
      leaf schema { type string; mandatory true; }
      leaf revision { type string; }
    }
    output {
      uses error-reporting;
      leaf source { type string; }
    }
  }
```

I.e. the `source` is a string.  Probably the fix here is to use `anydata` from YANG 1.1.

The other aspect that's a problem is the RPC protocol itself, which consists of messages with a length prefix.  We could instead use a strategy like HTTP's "chunked" transfer encoding.  Future work in any case.

wingo committed:
Fix use of undeclared variable (refactor residue)

wingo committed:
Fix default EOL handling style

wingo committed:
Fix error reporting in "listen" test

wingo committed:
Fix top-level scalar parsers to take stream as argument

wingo committed:
File streams query OS for appropriate block size

wingo committed:
Adapt "snabb config listen" to streaming parse

wingo committed:
Fix conversion of listen to streams

The socket isn't a stream, as we just use it to connect
and get a new fd.
